core/src/main/scala/akka/persistence/jdbc/state/scaladsl/JdbcDurableStateStore.scala

```diff
@@ -86,56 +87,90 @@ class JdbcDurableStateStore[A](
   def deleteObject(persistenceId: String): Future[Done] =
     db.run(queries._delete(persistenceId).map(_ => Done))
 
-  def currentChanges(tag: String, offset: Offset): Source[DurableStateChange[A], NotUsed] = offset match {
-    case NoOffset      => makeSourceFromStateQuery(tag, None)
-    case APSequence(l) => makeSourceFromStateQuery(tag, Some(l))
-    case _             => ??? // should not reach here
+  def currentChanges(tag: String, offset: Offset): Source[DurableStateChange[A], NotUsed] =
+    currentChangesByTag(tag, offset.value)
+
+  def changes(tag: String, offset: Offset): Source[DurableStateChange[A], NotUsed] =
+    _changesByTag(tag, offset.value, terminateAfterOffset = None)
+
+  private def currentChangesByTag(tag: String, offset: Long): Source[DurableStateChange[A], NotUsed] = {
+    Source
+      .futureSource(maxJournalSequence().map { maxOrderingInDb =>
+        _changesByTag(tag, offset, terminateAfterOffset = Some(maxOrderingInDb))
+      })
+      .mapMaterializedValue(_ => NotUsed)
+  }
+
+  private def _currentChangesByTag(
+      tag: String,
+      from: Long,
+      batchSize: Long,
+      queryUntil: MaxOffset): Source[DurableStateChange[A], NotUsed] = {
+    if (queryUntil.value < from.value) Source.empty
+    else changesByTag(tag, from, queryUntil.value, batchSize).mapAsync(1)(Future.fromTry)
+  }
+
+  private def changesByTag(
+      tag: String,
+      offset: Long,
+      maxOffset: Long,
+      max: Long): Source[Try[DurableStateChange[A]], NotUsed] = {
+    Source
+      .fromPublisher(db.stream(queries.changesByTag((tag, offset, maxOffset, max)).result))
+      .map(toDurableStateChange)
   }
 
-  def changes(tag: String, offset: Offset): Source[DurableStateChange[A], NotUsed] = {
+  // checks if the last n elements of a list are equal
+  private def equalLastNElements(coll: List[Long], n: Int): Boolean = {
+    if (coll.isEmpty) false
+    else if (coll.size < n) false
+    else {
+      val res = coll.slice(coll.length - n, coll.length)
+      res.forall(_ == res.head)
+    }
+  }
 
-    // a collection to store the start offsets for every iteration
-    val lastStartOffsets = collection.mutable.ListBuffer.empty[Long]
+  private def _changesByTag(
+      tag: String,
+      offset: Long,
+      terminateAfterOffset: Option[Long]): Source[DurableStateChange[A], NotUsed] = {
+    val batchSize = durableStateConfig.batchSize
+    val startingOffsets = collection.mutable.ListBuffer.empty[Long]
```

Comment:
This is used in a future callback (on other threads) so it is not safe to share a mutable structure like this. Would need to be an immutable collection passed as unfold state or a thread safe collection of some sort.

Reply:
Now changed .. immutability all the way ..